File: apps/web/src/lib/pricing/quotes.ts
import type { AllowedAsset } from "@/lib/assets/types";
import type { InvoiceCurrencyCode } from "@/lib/invoices/currencies";
import {
  ASSET_TO_COINGECKO,
  DEFAULT_INVOICE_QUOTE_TTL_MINUTES,
  DEFAULT_SLIPPAGE_BPS,
  RATE_CACHE_TTL_MS,
  STABLECOIN_FIAT_MAP,
} from "@/constants/pricing/quotes";
import {
  getCoinGeckoFiatCode,
} from "@/lib/invoices/currencies";
import { normalizeStellarAmount } from "@/lib/stellar/amount";

export { DEFAULT_SLIPPAGE_BPS };

type RateCacheEntry = {
  expiresAt: number;
  rates: Record<string, number>;
};

const rateCache = new Map<string, RateCacheEntry>();

function getQuoteTtlMinutes() {
  const parsed = Number(
    process.env.INVOICE_QUOTE_TTL_MINUTES ??
      String(DEFAULT_INVOICE_QUOTE_TTL_MINUTES)
  );
  return Number.isFinite(parsed) && parsed > 0
    ? parsed
    : DEFAULT_INVOICE_QUOTE_TTL_MINUTES;
}

function getSettlementFiatCode(assetCode: string): InvoiceCurrencyCode {
  return STABLECOIN_FIAT_MAP[assetCode] ?? "USD";
}

async function getFiatPerUsd(fiatCurrency: InvoiceCurrencyCode) {
  if (fiatCurrency === "USD") {
    return 1;
  }

  return fetchFiatPerUsd(fiatCurrency);
}

async function getInvoiceUsdValue(
  amount: string,
  currency: InvoiceCurrencyCode
) {
  const numeric = Number(amount);

  if (!Number.isFinite(numeric) || numeric <= 0) {
    throw new Error("Invalid invoice amount for quote");
  }

  const fiatPerUsd = await getFiatPerUsd(currency);
  return numeric / fiatPerUsd;
}

function assertQuoteUsdValue(
  invoiceUsdValue: number,
  paidAmount: string,
  paidAssetUsdPrice: number
) {
  const paidNumeric = Number(paidAmount);

  if (!Number.isFinite(paidNumeric) || paidNumeric <= 0) {
    throw new Error("Invalid quoted paid amount");
  }

  const paidUsdValue = paidNumeric * paidAssetUsdPrice;
  const ratio = paidUsdValue / invoiceUsdValue;

  // Reject quotes that are wildly off — catches inverted rates and bad price data.
  if (ratio < 0.5 || ratio > 2) {
    throw new Error(
      "Unable to build a reliable quote for this currency pair. Please try again."
    );
  }
}

async function fetchFiatPerUsd(fiatCurrency: InvoiceCurrencyCode) {
  if (fiatCurrency === "USD") {
    return 1;
  }

  const cacheKey = `fiat:${fiatCurrency}`;
  const cached = rateCache.get(cacheKey);

  if (cached && cached.expiresAt > Date.now()) {
    return cached.rates.usd;
  }

  const vs = getCoinGeckoFiatCode(fiatCurrency);
  const url = new URL("https://api.coingecko.com/api/v3/simple/price");
  url.searchParams.set("ids", "usd-coin");
  url.searchParams.set("vs_currencies", vs);

  const headers: HeadersInit = { Accept: "application/json" };
  const apiKey = process.env.COINGECKO_API_KEY?.trim();

  if (apiKey) {
    headers["x-cg-demo-api-key"] = apiKey;
  }

  const response = await fetch(url, { headers, next: { revalidate: 60 } });

  if (!response.ok) {
    throw new Error("Unable to fetch fiat exchange rates");
  }

  const data = (await response.json()) as Record<string, Record<string, number>>;
  const fiatPerUsd = data["usd-coin"]?.[vs];

  if (!fiatPerUsd || fiatPerUsd <= 0) {
    throw new Error(`Exchange rate unavailable for ${fiatCurrency}`);
  }

  rateCache.set(cacheKey, {
    expiresAt: Date.now() + RATE_CACHE_TTL_MS,
    rates: { usd: fiatPerUsd },
  });

  return fiatPerUsd;
}

async function convertFiatAmount(
  amount: string,
  from: InvoiceCurrencyCode,
  to: InvoiceCurrencyCode
) {
  if (from === to) {
    return amount;
  }

  const numeric = Number(amount);

  if (!Number.isFinite(numeric) || numeric <= 0) {
    throw new Error("Invalid amount for conversion");
  }

  const fromPerUsd = await getFiatPerUsd(from);
  const toPerUsd = await getFiatPerUsd(to);
  const usdAmount = numeric / fromPerUsd;
  const converted = usdAmount * toPerUsd;

  return converted.toFixed(7);
}

async function fetchAssetPriceInFiat(
  assetCode: string,
  fiatCurrency: InvoiceCurrencyCode
) {
  const pegged = STABLECOIN_FIAT_MAP[assetCode];

  if (pegged) {
    const peggedPerUsd = await getFiatPerUsd(pegged as InvoiceCurrencyCode);
    const targetPerUsd = await getFiatPerUsd(fiatCurrency);

    // Price of 1 stablecoin unit in the invoice fiat (e.g. USDC in IDR ≈ 16,000).
    const price = targetPerUsd / peggedPerUsd;

    if (!Number.isFinite(price) || price <= 0) {
      throw new Error(`Exchange rate unavailable for ${assetCode}`);
    }

    if (price < 0.01) {
      throw new Error(`Exchange rate out of acceptable range for ${assetCode}`);
    }

    return price;
  }

  const cacheKey = `asset:${assetCode}:${fiatCurrency}`;
  const cached = rateCache.get(cacheKey);

  if (cached && cached.expiresAt > Date.now()) {
    return cached.rates.price;
  }

  const coinId = ASSET_TO_COINGECKO[assetCode];

  if (!coinId) {
    throw new Error(`Pricing is not available for ${assetCode}`);
  }

  const vs = getCoinGeckoFiatCode(fiatCurrency);
  const url = new URL("https://api.coingecko.com/api/v3/simple/price");
  url.searchParams.set("ids", coinId);
  url.searchParams.set("vs_currencies", vs);

  const headers: HeadersInit = { Accept: "application/json" };
  const apiKey = process.env.COINGECKO_API_KEY?.trim();

  if (apiKey) {
    headers["x-cg-demo-api-key"] = apiKey;
  }

  const response = await fetch(url, { headers, next: { revalidate: 60 } });

  if (!response.ok) {
    throw new Error("Unable to fetch asset prices");
  }

  const data = (await response.json()) as Record<string, Record<string, number>>;
  const price = data[coinId]?.[vs];

  if (!price || price <= 0) {
    throw new Error(`Price unavailable for ${assetCode}`);
  }

  rateCache.set(cacheKey, {
    expiresAt: Date.now() + RATE_CACHE_TTL_MS,
    rates: { price },
  });

  return price;
}

export async function getAssetToFiatRate(
  assetCode: string,
  fiatCurrency: InvoiceCurrencyCode
) {
  return fetchAssetPriceInFiat(assetCode, fiatCurrency);
}

export function applySendMaxBuffer(amount: string, slippageBps = DEFAULT_SLIPPAGE_BPS) {
  const numeric = Number(amount);

  if (!Number.isFinite(numeric) || numeric <= 0) {
    throw new Error("Invalid amount for send max buffer");
  }

  const buffered = numeric * (1 + slippageBps / 10_000);
  return normalizeStellarAmount(buffered.toFixed(7));
}

export function assetsMatch(
  left: AllowedAsset,
  right: AllowedAsset
) {
  return (
    left.asset_code === right.asset_code &&
    (left.issuer_address?.trim() || null) === (right.issuer_address?.trim() || null)
  );
}

export async function buildPaymentQuote(input: {
  pricingAmount: string;
  pricingCurrency: InvoiceCurrencyCode;
  paidAsset: AllowedAsset;
  settlementAsset: AllowedAsset;
}) {
  const invoiceTotal = Number(input.pricingAmount);

  if (!Number.isFinite(invoiceTotal) || invoiceTotal <= 0) {
    throw new Error("Invalid invoice amount for quote");
  }

  const assetPriceInFiat = await fetchAssetPriceInFiat(
    input.paidAsset.asset_code,
    input.pricingCurrency
  );

  const paidAmountRaw = invoiceTotal / assetPriceInFiat;
  const paidAmount = normalizeStellarAmount(paidAmountRaw.toFixed(7));

  const invoiceUsdValue = await getInvoiceUsdValue(
    input.pricingAmount,
    input.pricingCurrency
  );
  const paidAssetUsdPrice = await fetchAssetPriceInFiat(
    input.paidAsset.asset_code,
    "USD"
  );
  assertQuoteUsdValue(invoiceUsdValue, paidAmount, paidAssetUsdPrice);

  const settlementFiat = getSettlementFiatCode(input.settlementAsset.asset_code);
  const settlementFiatAmount = await convertFiatAmount(
    input.pricingAmount,
    input.pricingCurrency,
    settlementFiat
  );
  const settlementAmount = normalizeStellarAmount(settlementFiatAmount);

  const paidNumeric = Number(paidAmount);
  const settlementNumeric = Number(settlementAmount);
  const settlementQuoteRate =
    paidNumeric > 0
      ? (settlementNumeric / paidNumeric).toFixed(7)
      : "0";

  const expiresAt = new Date(Date.now() + getQuoteTtlMinutes() * 60 * 1000);

  return {
    pricing_amount: input.pricingAmount,
    pricing_currency: input.pricingCurrency,
    paid_asset: input.paidAsset,
    paid_amount: paidAmount,
    settlement_asset: input.settlementAsset,
    settlement_amount: settlementAmount,
    rate: assetPriceInFiat.toString(),
    settlement_quote_rate: settlementQuoteRate,
    requires_path_payment: !assetsMatch(input.paidAsset, input.settlementAsset),
    expires_at: expiresAt,
  };
}

export function isQuoteExpired(expiresAt: Date | null | undefined) {
  return Boolean(expiresAt && expiresAt.getTime() <= Date.now());
}

export function amountsWithinSlippage(
  expected: string,
  actual: string,
  slippageBps = DEFAULT_SLIPPAGE_BPS
) {
  const expectedNum = Number(expected);
  const actualNum = Number(actual);

  if (!Number.isFinite(expectedNum) || !Number.isFinite(actualNum)) {
    return false;
  }

  if (expectedNum === 0) {
    return actualNum === 0;
  }

  const diff = Math.abs(actualNum - expectedNum) / expectedNum;
  return diff <= slippageBps / 10_000;
}

File: apps/api/internal/pricing/quotes.go
// Package pricing ports apps/web/src/lib/pricing/quotes.ts.
package pricing

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	DefaultSlippageBPS           = 50
	RateCacheTTL                 = 60 * time.Second
	DefaultInvoiceQuoteTTLMinutes = 15
)

var (
	stablecoinFiatMap = map[string]string{
		"USDC": "USD",
		"PYUSD": "USD",
		"EURC": "EUR",
		"AUDD": "AUD",
		"NGNC": "USD",
	}
	assetToCoinGecko = map[string]string{
		"XLM":  "stellar",
		"USDC": "usd-coin",
		"EURC": "euro-coin",
		"PYUSD": "paypal-usd",
		"AUDD": "novatti-australian-digital-dollar",
	}
)

type AllowedAsset struct {
	AssetCode     string  `json:"asset_code"`
	IssuerAddress *string `json:"issuer_address"`
}

type Quote struct {
	PricingAmount       string       `json:"pricing_amount"`
	PricingCurrency     string       `json:"pricing_currency"`
	PaidAsset           AllowedAsset `json:"paid_asset"`
	PaidAmount          string       `json:"paid_amount"`
	SettlementAsset     AllowedAsset `json:"settlement_asset"`
	SettlementAmount    string       `json:"settlement_amount"`
	Rate                string       `json:"rate"`
	SettlementQuoteRate string       `json:"settlement_quote_rate"`
	RequiresPathPayment bool         `json:"requires_path_payment"`
	ExpiresAt           time.Time    `json:"expires_at"`
}

type rateCacheEntry struct {
	expiresAt time.Time
	rates     map[string]float64
}

var (
	rateCache   = map[string]rateCacheEntry{}
	rateCacheMu sync.Mutex
)

func quoteTTLMinutes(configured int) int {
	if configured > 0 {
		return configured
	}
	return DefaultInvoiceQuoteTTLMinutes
}

func NormalizeStellarAmount(amount string) string {
	parts := strings.Split(strings.TrimSpace(amount), ".")
	whole := "0"
	frac := ""
	if len(parts) > 0 && parts[0] != "" {
		whole = parts[0]
	}
	if len(parts) > 1 {
		frac = parts[1]
	}
	frac = (frac + "0000000")[:7]
	return whole + "." + frac
}

func AssetsMatch(left, right AllowedAsset) bool {
	leftIssuer := ""
	rightIssuer := ""
	if left.IssuerAddress != nil {
		leftIssuer = strings.TrimSpace(*left.IssuerAddress)
	}
	if right.IssuerAddress != nil {
		rightIssuer = strings.TrimSpace(*right.IssuerAddress)
	}
	return left.AssetCode == right.AssetCode && leftIssuer == rightIssuer
}

func IsQuoteExpired(expiresAt *time.Time) bool {
	return expiresAt != nil && !expiresAt.After(time.Now())
}

func AmountsWithinSlippage(expected, actual string, slippageBPS int) bool {
	if slippageBPS <= 0 {
		slippageBPS = DefaultSlippageBPS
	}
	exp, err1 := strconv.ParseFloat(expected, 64)
	act, err2 := strconv.ParseFloat(actual, 64)
	if err1 != nil || err2 != nil {
		return false
	}
	if exp == 0 {
		return act == 0
	}
	diff := (act - exp) / exp
	if diff < 0 {
		diff = -diff
	}
	return diff <= float64(slippageBPS)/10_000
}

func ApplySendMaxBuffer(amount string, slippageBPS int) (string, error) {
	if slippageBPS <= 0 {
		slippageBPS = DefaultSlippageBPS
	}
	numeric, err := strconv.ParseFloat(amount, 64)
	if err != nil || numeric <= 0 {
		return "", fmt.Errorf("Invalid amount for send max buffer")
	}
	buffered := numeric * (1 + float64(slippageBPS)/10_000)
	return NormalizeStellarAmount(fmt.Sprintf("%.7f", buffered)), nil
}

func BuildPaymentQuote(pricingAmount, pricingCurrency string, paidAsset, settlementAsset AllowedAsset, ttlMinutes int) (*Quote, error) {
	invoiceTotal, err := strconv.ParseFloat(pricingAmount, 64)
	if err != nil || invoiceTotal <= 0 {
		return nil, fmt.Errorf("Invalid invoice amount for quote")
	}

	assetPriceInFiat, err := fetchAssetPriceInFiat(paidAsset.AssetCode, pricingCurrency)
	if err != nil {
		return nil, err
	}
	paidAmount := NormalizeStellarAmount(fmt.Sprintf("%.7f", invoiceTotal/assetPriceInFiat))

	invoiceUSD, err := getInvoiceUSDValue(pricingAmount, pricingCurrency)
	if err != nil {
		return nil, err
	}
	paidAssetUSD, err := fetchAssetPriceInFiat(paidAsset.AssetCode, "USD")
	if err != nil {
		return nil, err
	}
	if err := assertQuoteUSDValue(invoiceUSD, paidAmount, paidAssetUSD); err != nil {
		return nil, err
	}

	settlementAmount := paidAmount
	if !AssetsMatch(paidAsset, settlementAsset) {
		settlementPrice, err := fetchAssetPriceInFiat(settlementAsset.AssetCode, pricingCurrency)
		if err != nil {
			return nil, err
		}
		settlementAmount = NormalizeStellarAmount(fmt.Sprintf("%.7f", invoiceTotal/settlementPrice))
	}

	paidNumeric, _ := strconv.ParseFloat(paidAmount, 64)
	settlementNumeric, _ := strconv.ParseFloat(settlementAmount, 64)
	settlementQuoteRate := "0"
	if paidNumeric > 0 {
		settlementQuoteRate = fmt.Sprintf("%.7f", settlementNumeric/paidNumeric)
	}

	return &Quote{
		PricingAmount:       pricingAmount,
		PricingCurrency:     pricingCurrency,
		PaidAsset:           paidAsset,
		PaidAmount:          paidAmount,
		SettlementAsset:     settlementAsset,
		SettlementAmount:    settlementAmount,
		Rate:                strconv.FormatFloat(assetPriceInFiat, 'f', -1, 64),
		SettlementQuoteRate: settlementQuoteRate,
		RequiresPathPayment: !AssetsMatch(paidAsset, settlementAsset),
		ExpiresAt:           time.Now().Add(time.Duration(quoteTTLMinutes(ttlMinutes)) * time.Minute),
	}, nil
}

func assertQuoteUSDValue(invoiceUSD float64, paidAmount string, paidAssetUSD float64) error {
	paidNumeric, err := strconv.ParseFloat(paidAmount, 64)
	if err != nil || paidNumeric <= 0 {
		return fmt.Errorf("Invalid quoted paid amount")
	}
	ratio := (paidNumeric * paidAssetUSD) / invoiceUSD
	if ratio < 0.5 || ratio > 2 {
		return fmt.Errorf("Unable to build a reliable quote for this currency pair. Please try again.")
	}
	return nil
}

func getInvoiceUSDValue(amount, currency string) (float64, error) {
	numeric, err := strconv.ParseFloat(amount, 64)
	if err != nil || numeric <= 0 {
		return 0, fmt.Errorf("Invalid invoice amount for quote")
	}
	fiatPerUSD, err := getFiatPerUSD(currency)
	if err != nil {
		return 0, err
	}
	return numeric / fiatPerUSD, nil
}

func getFiatPerUSD(fiatCurrency string) (float64, error) {
	if fiatCurrency == "USD" {
		return 1, nil
	}
	return fetchFiatPerUSD(fiatCurrency)
}

func fetchFiatPerUSD(fiatCurrency string) (float64, error) {
	cacheKey := "fiat:" + fiatCurrency
	rateCacheMu.Lock()
	if cached, ok := rateCache[cacheKey]; ok && cached.expiresAt.After(time.Now()) {
		rateCacheMu.Unlock()
		return cached.rates["usd"], nil
	}
	rateCacheMu.Unlock()

	vs := strings.ToLower(fiatCurrency)
	u, _ := url.Parse("https://api.coingecko.com/api/v3/simple/price")
	q := u.Query()
	q.Set("ids", "usd-coin")
	q.Set("vs_currencies", vs)
	u.RawQuery = q.Encode()

	data, err := coinGeckoGet(u.String())
	if err != nil {
		return 0, fmt.Errorf("Unable to fetch fiat exchange rates")
	}
	fiatPerUSD, ok := data["usd-coin"][vs]
	if !ok || fiatPerUSD <= 0 {
		return 0, fmt.Errorf("Exchange rate unavailable for %s", fiatCurrency)
	}

	rateCacheMu.Lock()
	rateCache[cacheKey] = rateCacheEntry{
		expiresAt: time.Now().Add(RateCacheTTL),
		rates:     map[string]float64{"usd": fiatPerUSD},
	}
	rateCacheMu.Unlock()
	return fiatPerUSD, nil
}

func fetchAssetPriceInFiat(assetCode, fiatCurrency string) (float64, error) {
	if pegged, ok := stablecoinFiatMap[assetCode]; ok {
		peggedPerUSD, err := getFiatPerUSD(pegged)
		if err != nil {
			return 0, err
		}
		targetPerUSD, err := getFiatPerUSD(fiatCurrency)
		if err != nil {
			return 0, err
		}
		price := targetPerUSD / peggedPerUSD
		if price <= 0 {
			return 0, fmt.Errorf("Exchange rate unavailable for %s", assetCode)
		}
		if price < 0.01 {
			return 0, fmt.Errorf("Exchange rate out of acceptable range for %s", assetCode)
		}
		return price, nil
	}

	cacheKey := "asset:" + assetCode + ":" + fiatCurrency
	rateCacheMu.Lock()
	if cached, ok := rateCache[cacheKey]; ok && cached.expiresAt.After(time.Now()) {
		rateCacheMu.Unlock()
		return cached.rates["price"], nil
	}
	rateCacheMu.Unlock()

	coinID, ok := assetToCoinGecko[assetCode]
	if !ok {
		return 0, fmt.Errorf("Pricing is not available for %s", assetCode)
	}
	vs := strings.ToLower(fiatCurrency)
	u, _ := url.Parse("https://api.coingecko.com/api/v3/simple/price")
	q := u.Query()
	q.Set("ids", coinID)
	q.Set("vs_currencies", vs)
	u.RawQuery = q.Encode()

	data, err := coinGeckoGet(u.String())
	if err != nil {
		return 0, fmt.Errorf("Unable to fetch asset prices")
	}
	price, ok := data[coinID][vs]
	if !ok || price <= 0 {
		return 0, fmt.Errorf("Price unavailable for %s", assetCode)
	}

	rateCacheMu.Lock()
	rateCache[cacheKey] = rateCacheEntry{
		expiresAt: time.Now().Add(RateCacheTTL),
		rates:     map[string]float64{"price": price},
	}
	rateCacheMu.Unlock()
	return price, nil
}

func coinGeckoGet(rawURL string) (map[string]map[string]float64, error) {
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	if key := strings.TrimSpace(os.Getenv("COINGECKO_API_KEY")); key != "" {
		req.Header.Set("x-cg-demo-api-key", key)
	}
	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("coingecko status %d", resp.StatusCode)
	}
	var data map[string]map[string]float64
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, err
	}
	return data, nil
}
